### PR TITLE
Adds Collection Type Participants functionality

### LINKED ISSUE #1345: Collection Types: Create participants partial of the new/edit form
### Descriptive summary

Add code to implement the Participants tab partial for the new/edit form views for collection types.  This partial is shared by both the new and edit form views.

### Rationale

See model issue for rationale.

### Implementation

**Files:**
* app/views/hyrax/admin/collection_types/_form_participants.html.erb

#### Expected user interactions

Create the UI view for adding participants for a collection type.

The form fields allow the user to select a group or user and apply a participant role to that group/user.  The roles for collection types includes...
- [ ] Collection Manager
- [ ] Collection Creator

The form will also include a listing of groups/users assigned to each participant role.

Exact text for the page is in the mockup below.

#### SPECIAL NOTE on Admin Sets:

For the initial implementaiton, the Admin Set underlying code will not change.  But to a user, the Admin Set will appear to be a collection type.

To make the illusion happen...
* as part of the hyrax generation process, a collection type will be created for Administrative Sets and the id will be recorded in a variable that can be checked by this UI (see issue #1342 for Admin Set Generator)
* the initial values for these settings are set by the generator to be the set of admins
* just like other types, these values can be changed at any time

### Mockup

new-collection-type-participants-tab

### Related work

| Collection Types feature | Issue |
| ------------------------ | ----- |
| Admin Set Generator      | #1342 |
| User Collection Generator | #1485  |
| Routes                   | #1347 |
| Controller               | #1341 |
| Model                    | #1343 |
| View - Index             | #1349 |
| View -  New/Edit         | #1348 |
| Partial -  Description   | #1344 |
| Partial -  Settings      | #1346 |
| Partial -  Participants  | this issue |

### PR DESCRIPTION
**Pull request by kdid**
Partially fixes #1345

Fixes #1345 - Collection Types Participants Partial

**Description**
Adds functionality to add/remove participants to/from a collection type.
  - Form to add users or groups to a collection type either as creators or managers
  - lists current participants of the collection type, with button to remove participant

**Changes in this pull request**
- Migration for database table collection_type_participants that stores participant access info
- Adds `Admin::CollectionTypeParticipantsController`, `CollectionTypeParticipantForm`, `CollectionTypeParticipant` model associated tests
- Adds`/views/hyrax/admin/collection_types/_form_participants.html.erb` and `/views/hyrax/admin/collection_types/_form_participants_table.html.erb`

**Remaining Issues:**
- The `/views/hyrax/admin/collection_types/_form.html.erb` view needs to be rearranged. The `_form_participants.html.erb` partial cannot nest within the other form but that form is spanning the other two tabs. I moved the participants partial under the submit button as a temporary measure during development, but it needs permanent solution. Maybe if the form.persisted? is true the first tab could become the "Metadata and Discovery" and contain the other two partials? That way the form could live within the first tab pane and not span them all.

@samvera/hyrax-code-reviewers

### CODE REVIEW

File: app/controllers/hyrax/admin/collection_type_participants_controller.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    def create
+      @collection_type_participant = Hyrax::CollectionTypeParticipant.new(collection_type_participant_params)
+      if @collection_type_participant.save
+        redirect_to(
+          edit_admin_collection_type_path(@collection_type_participant.hyrax_collection_type_id, anchor: 'participants'),
+          notice: I18n.t('update_notice', scope: 'hyrax.admin.collection_types.form_participants')
+        )
+      else
+        redirect_to(
+          hyrax.edit_admin_collection_type_path(@collection_type_participant.hyrax_collection_type_id, anchor: 'participants'),
```
Comment (jeremyf) [inline app/controllers/hyrax/admin/collection_type_participants_controller.rb]: Why is the redirect route prefixed with `hyrax` but not on line 14?

File: spec/controllers/hyrax/admin/collection_type_participants_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+    before do
+      allow(controller.current_ability).to receive(:can?).with(any_args).and_return(false)
+      sign_in user
+    end
+
+    describe "#create" do
+      it "returns http redirect" do
+        post :create
+        expect(response).to have_http_status(:redirect)
```
Comment (jeremyf) [inline spec/controllers/hyrax/admin/collection_type_participants_controller_spec.rb]: This spec needs to assert that the `#create` action failed. Both success and failure results in a redirect. What is the differentiation?

File: spec/controllers/hyrax/admin/collection_type_participants_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+      end
+    end
+
+    describe "#destroy" do
+      it "returns http redirect" do
+        delete :destroy, params: { id: :id }
+        expect(response).to have_http_status(:redirect)
```
Comment (jeremyf) [inline spec/controllers/hyrax/admin/collection_type_participants_controller_spec.rb]: This spec needs to assert that the `#destroy` action failed. Both success and failure results in a redirect. What is the differentiation?

File: spec/controllers/hyrax/admin/collection_type_participants_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,104 @@
+RSpec.describe Hyrax::Admin::CollectionTypeParticipantsController, type: :controller do
+  context 'anonymous user' do
+    describe '#create' do
+      it 'returns http redirect' do
+        post :create
+        expect(response).to have_http_status(:redirect)
```
Comment (jeremyf) [inline spec/controllers/hyrax/admin/collection_type_participants_controller_spec.rb]: This spec needs to assert that the `#create` action failed. Both success and failure results in a redirect. What is the differentiation?

File: spec/controllers/hyrax/admin/collection_type_participants_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+      end
+    end
+
+    describe '#destroy' do
+      it 'returns http redirect' do
+        delete :destroy, params: { id: :id }
+        expect(response).to have_http_status(:redirect)
```
Comment (jeremyf) [inline spec/controllers/hyrax/admin/collection_type_participants_controller_spec.rb]: This spec needs to assert that the `#destroy` action failed. Both success and failure results in a redirect. What is the differentiation?

**Comment by kdid** [comment]
@jeremyf - I made the changes you suggested if you want to take another look. Thanks.

**Comment by kdid** [comment]
Closing. Opened New PR here: https://github.com/samvera/hyrax/pull/1546
#1546

**Pull request closed without merging**